Support dynamo tracing collectives with processgroup arg

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102232
* #102412
* __->__ #102222
* #102379

Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

wconstab committed:
Support dynamo tracing collectives with processgroup arg

Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

[ghstack-poisoned]

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wanchaol commented on torch/_dynamo/variables/torch.py:
```diff
@@ -128,9 +128,14 @@ class TorchVariable(VariableTracker):
+                value = tensor_dunder_fns_remap[value]
+        except TypeError as e:
+            # is it cleaner to somehow exclude unhashable types from the `in` check above?
+            if "unhashable type" not in e.args[0]:
```
for my learning: is the bytecode directly generate "unhashable type" for process group so that we need to handle it here when we calling sth like `torch.distributed._functional_collective.all_gather_tnesor`?

wconstab replied:
no, the issue here is that someone wrote code that assumes 'value' will be a function and can be looked up in a dict of functions to remap.  But value isn't always a function and so i had to either check if its a function or check if its hashable to be able to look it up safely.

wanchaol commented on torch/_dynamo/variables/torch.py:
```diff
@@ -494,6 +499,15 @@ def get_state_from_generator():
                 tx, [args[0], result], {}
             )
+        elif (
+            self.value == torch.distributed.get_process_group_ranks
```
I assume we capture into the functional collectives so that we get these two calls and bake the pg -> ranks to the IR?

voznesenskym replied:
Yes

wanchaol reviewed (approved):
lgtm, some questions.

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -128,9 +128,14 @@ class TorchVariable(VariableTracker):
+                value = tensor_dunder_fns_remap[value]
+        except TypeError as e:
+            # is it cleaner to somehow exclude unhashable types from the `in` check above?
+            if "unhashable type" not in e.args[0]:
```
Don't do this, parsing strings from exceptions is weird. Use https://docs.python.org/3/library/collections.abc.html#collections.abc.Hashable

wconstab replied:
i think i actually want to insert an `inspect.isfunction(value)` before checking the remap.  is that a better way to do this?

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -494,6 +499,15 @@ def get_state_from_generator():
+            self.value == torch.distributed.get_process_group_ranks
+            or self.value == torch.distributed.distributed_c10d._get_group_tag
+        ):
+            # guard on this 'pg'
```
This isn't actually guarding tho?

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -510,3 +510,28 @@ def odict_getitem(self, tx, key):
         )
+
+
+class ProcessGroupVariable(UserDefinedObjectVariable):
```
not sold on this being UserDefinedObjectVariable but let's keep it for now. I was thinking of actually just having it be a top level VT since we don't want to inline it, but rather, we want to intercept certain functions and bake out consts (rank, size), and skip the rest.

voznesenskym commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -510,3 +510,28 @@ def odict_getitem(self, tx, key):
         )
+
+
+class ProcessGroupVariable(UserDefinedObjectVariable):
+    """
+    We don't want a ProcessGroup object to end up in our output graph.
+
+    But it's common for dynamo to intercept a PG that is then used to get info like
+    rank() or world_size(), as well as passed to utility functions in distributed_c10d
+    which desugar it into plain types like a ranklist and tag.
+
+    For convenience and proper guarding, we construct a variable type.
+
+    TODO: implement+test common/safe methods such as get_rank
+    TODO: make it possible to use ProcessGroupVariable as input to simple functions
+          like _expand_group without dynamo complaining about making a proxy for it.
+          It is not a tensor-like type, and we don't want a proxy- but dynamo assumes
+          torch library functions are dealing with tensor-like types and would have proxies
+          for their args.
+    """
+
+    def __init__(self, value, **kwargs):
+        super().__init__(value, **kwargs)
+
+    def as_python_constant(self):
+        return self.value
```
Rewritten from the top-down fsdp tracing experimentation:

```suggestion
class ProcessGroupVariable(UserDefinedObjectVariable):
    """
    We don't want a ProcessGroup object to end up in our output graph.
    But it's common for dynamo to intercept a PG that is then used to get info like
    rank() or world_size(), as well as passed to utility functions in distributed_c10d
    which desugar it into plain types like a ranklist and tag.
    For convenience and proper guarding, we construct a variable type.
    TODO: implement+test common/safe methods such as get_rank
    TODO: make it possible to use ProcessGroupVariable as input to simple functions
          like _expand_group without dynamo complaining about making a proxy for it.
          It is not a tensor-like type, and we don't want a proxy- but dynamo assumes
          torch library functions are dealing with tensor-like types and would have proxies
          for their args.
    """

    def __init__(self, value, **kwargs):
        super().__init__(value, **kwargs)

    def as_python_constant(self):
        return self.value
    
    def call_function(
        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
    ) -> "VariableTracker":
        print("PG INVOKED", self, [type(arg) for arg in args])
        return super().call_function(tx, args, kwargs)

    def call_method(
        self,
        tx,
        name,
        args: "List[VariableTracker]",
        kwargs: "Dict[str, VariableTracker]",
    ) -> "VariableTracker":
        if name == "rank":
            return variables.ConstantVariable(self.value.rank)
        if name == "size":
            return variables.ConstantVariable(self.value.rank)
        return super().call_method(tx, name, args, kwargs)
        
    def var_getattr(self, tx, name):
        if name in ["rank", "size"]:
            return variables.LambdaVariable(
                lambda *args, **kwargs: self.call_method(tx, name, args, kwargs)
            ).add_options(self)
        return super().var_getattr(tx, name)
```

wconstab replied:
Ahh LambdaVariable is great. I didn't know we had that.

Also, I don't understand why we want `call_function` - is it ever valid to write user code like `my_processgroup()`?  (Maybe it is, but i just didn't think that was a thing)

voznesenskym replied:
You can drop the `call_function` - this is just here for debugging. It will automatically go to super on its own.

wconstab replied:
cool. yea i dropped it

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -510,3 +510,52 @@ def odict_getitem(self, tx, key):
+        if name == "size":
+            return variables.ConstantVariable(self.value.size())
+
+        # TODO should this just raise unimplemented?
+        return super().call_method(tx, name, args, kwargs)
```
No, because you are a UserDefinedObjectVariable.

If you want to raise unimplemented, the idiomatic way is to make this a top-level variable-tracker and still invoke super, which will do it for you.

wconstab replied:
yea makes sense. i agree, we should move to being a top-level VT.

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab commented:
@pytorchhbot merge

wconstab commented:
@pytorchbot merge

wconstab committed:
Update on "Support dynamo tracing collectives with processgroup arg"


Previously, other types of rank descriptors worked but pg
caused dynamo to break down when tracing the internal func that converts
pg to rank list.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
